### PR TITLE
Push pending network backend changes into 2-11 base

### PR DESCRIPTION
**Pull request by Salanto**
This marks the beginning of the end of courtroom.cpp as we know it.
The hydra will be slain or I will go down trying.

### CODE REVIEW

**Comment by stonedDiscord** [comment]
157 files changed

File: attorney_online/include/aolayer.h [hunk 1]
```diff
@@ -153,6 +154,8 @@ class AOLayer : public QLabel {
   QMutex mutex;
   QWaitCondition frameAdded;
 
+  Options options;
```
Comment (stonedDiscord) [inline attorney_online/include/aolayer.h:157]: what does this do
options could be anything

File: attorney_online/include/datatypes.h [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+    AUTH,       //!< Authentication packet - controls guard button
+    ZZ,         //!< Modcall
+    KK,         //!< Kicked
+    KB,         //!< Kicked and banned
+    BD,         //!< Banned
+    BB,         //!< Bessage Box
+    CHECK,      //!< Keep-alive packet
+    ST,         //!< Serverside subtheme selection
+    TI,         //!< Timers
+    JD          //!< Judge Controls
```
Comment (stonedDiscord) [inline attorney_online/include/datatypes.h:73]: good one

**Review by stonedDiscord — APPROVED** [review]

**Comment by stonedDiscord** [comment]
this is a good one, loving it so far

personally i don't like that you changed the entire directory structure for almost no reason, but i can look over that

**Comment by Salanto** [comment]
Reason being that the old structure made it very had to add QTest cleanly into the project structure without resorting to manual inclusion of source and header files or becoming very stringy on where stuff can be placed internally.

I already have to do that as AO2, unlike Akashi, does not bundle the app as a library to link against, which you can see in the partially finished datattype hander tests where the inclusion for MOC looks a bit odd.

Separating them out like this gives it the easiest structure to work with.

File: attorney_online/include/datatypes.h [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+          other_offset = {offset[0].toInt(), 0};
+        }
+        other_flip = toDataType<bool>(packet.at(OTHER_FLIP));
+        immediate = toDataType<bool>(packet.at(IMMEDIATE));
+      }
+      if (kfo_features) {
+        sfx_looping = toDataType<bool>(packet.at(LOOPING_SFX));
+        screenshake = toDataType<bool>(packet.at(SCREENSHAKE));
+        if (!packet.at(FRAME_SFX).isEmpty() &&
+            Options().networkedFrameSfxEnabled()) {
```
Comment (oldmud0) [inline attorney_online/include/datatypes.h:298]: this looks smelly to be accessing settings from a struct. For instance, when you're saving messages to a replay, and you have networked frame sfx disabled, you might want to preserve that message data instead of throwing it away just because a clientside option was set.

File: attorney_online/include/datatypes.h [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+{
+  return QVariant::fromValue(f_t).toString();
+}
+
+namespace AttorneyOnline {
+
+Q_NAMESPACE
+
+class DataTypes {
+  Q_GADGET
```
Comment (oldmud0) [inline attorney_online/include/datatypes.h:24]: i had to look this up. legit

File: attorney_online/include/datatypes.h [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+    LOGOUT = -1,  //!< Log-out. Hides the guard button.
+    LOGIN_FAIL,   //!< Unsuccessful log-in attempt.
+    LOGIN_SUCCESS //!< Successful log-in. Displays the guard button.
+  };
+  Q_ENUM(AUTHENTICATION)
+
+  /**
+   * @brief The TIMER_ACTION enum
+   *
+   *  @details Allan, please add details.
```
Comment (oldmud0) [inline attorney_online/include/datatypes.h:161]: who is allan

File: attorney_online/include/lobby.h [hunk 6]
```diff
@@ -35,15 +37,18 @@ class Lobby : public QMainWindow {
 
 private:
   AOApplication *ao_app;
-  NetworkManager *net_manager;
+  NetworkManager* net_manager;
+  std::shared_ptr<AttorneyOnline::Client>* client;
```
Comment (oldmud0) [inline attorney_online/include/lobby.h:41]: please don't use a pointer of a shared_ptr. It makes me cry.

**Review by oldmud0 — COMMENTED** [review]
This is legit. I couldn't have asked for anything else. Thank you. This is amazing

File: attorney_online/include/datatypes.h (hunk 5 above)
Comment (Salanto) [reply attorney_online/include/datatypes.h:161]: image

I admit, I got lazy and did not document these as briefly as "at glance" understanding of AOs network without resorting to opening network.md requires. This was supposed to be a reminder to do it, which I did not. 

TLDR : It turns out I am Allan.

File: attorney_online/include/datatypes.h (hunk 3 above)
Comment (Salanto) [reply attorney_online/include/datatypes.h:298]: Valid, I forgot demo files were a thing while writing that part and that I need to actually handle them at some point, but rn I am more concerned with getting the fundamentals more right than wrong than worry about AOs more than few crimes against sane codebases.
